### PR TITLE
[distill][phase0] Refactor distillation scaffold: Method/Trainer/Adapter + Wan DMD2 v2 entrypoint

### PR DESCRIPTION
**Pull request by alexzms**
## 1) Motivation
The current Wan DMD distillation training in FastVideo is heavily coupled inside `fastvideo/training/*distillation_pipeline*.py` and Wan-specific helpers. This makes it difficult to:
- Reuse the same distillation algorithm (DMD2 / self-forcing / …) across different models/pipelines.
- Introduce more roles beyond the current hard-coded set (teacher / student / critic / reward / refiner / …).
- Review and iterate safely, because training loop logic, algorithm logic, validation, and conditioning are intertwined.
- Avoid hidden dependencies (e.g., unconditional/negative conditioning being initialized only as a validation side effect).

```mermaid
flowchart LR
  Run["Run FastVideo distillation"] --> P1["WanDistillationPipeline<br/>(Wan + DMD2)"]
  Run --> P2["WanSelfForcingDistillationPipeline<br/>(Wan + Self-Forcing)"]
  Run --> P3["WanI2VDistillationPipeline<br/>(Wan + DMD2 I2V)"]

  P1 --> Base["DistillationPipeline (base)"]
  P2 --> Base
  P3 --> Base

  Base --> Coupling["Model type (Wan) + distill method<br/>are bundled inside pipelines<br/>(not independently swappable)"]

  classDef bad fill:#ffecec,stroke:#cc0000,stroke-width:2px
  class Coupling bad

```

This PR starts a refactor to decouple the distillation pipeline into clean layers (Trainer ↔ Method ↔ Adapter), enabling incremental, reviewable PRs rather than a single large rewrite.

## 2) What we can learn from FastGen (A framework that is particularly designed for training models using different distillation method)
FastGen has a few design choices that are particularly effective for distillation pipelines:
- A thin `Trainer` that only orchestrates iteration, grad accumulation, stepping/zeroing, validation, and checkpointing.
- Algorithm logic lives in the model/method interface (`single_train_step()`, `optimizers_schedulers_step()`, etc.), so multi-optimizer schedules are controlled by the algorithm rather than being baked into the trainer.
- A clear separation between “method/algorithm” and “network/backbone”, making it easy to apply the same method to different networks with minimal glue code.

```mermaid
flowchart LR
  %% --- 样式定义区 ---
  classDef infra fill:#e3f2fd,stroke:#1565c0,stroke-width:2px,rx:10,ry:10,color:#0d47a1,font-weight:bold
  classDef abstract fill:#f3e5f5,stroke:#8e24aa,stroke-width:2px,rx:10,ry:10,color:#4a148c,font-weight:bold
  classDef methodNode fill:#e8f5e9,stroke:#2e7d32,stroke-width:2px,rx:5,ry:5,color:#1b5e20
  classDef netNode fill:#ffe0b2,stroke:#e65100,stroke-width:2px,rx:5,ry:5,color:#e65100

  %% --- 主图结构 ---
  C["Config"]:::infra --> M["Method (FastGenModel)"]:::abstract
  C --> N["Network (FastGenNetwork)"]:::abstract

  T["Trainer"]:::infra -->|"train/run"| M

  subgraph Methods ["fastgen/methods/ (examples)"]
    %% 设置 fill:none 去掉背景框颜色，保留虚线边框
    style Methods fill:none,stroke:#aed581,stroke-width:2px,stroke-dasharray: 5 5,color:#33691e

    M --> DM["distribution_matching/..."]:::methodNode
    M --> KD["knowledge_distillation/..."]:::methodNode
    M --> CM["consistency_model/..."]:::methodNode
    M --> FT["fine_tuning/..."]:::methodNode
  end

  subgraph Networks ["fastgen/networks/ (examples)"]
    %% 设置 fill:none 去掉背景框颜色
    style Networks fill:none,stroke:#ffb74d,stroke-width:2px,stroke-dasharray: 5 5,color:#e65100

    N --> N1["Wan"]:::netNode
    N --> N2["CogVideoX"]:::netNode
    N --> Nx["..."]:::netNode
  end

```
## 3) Proposed FastVideo architecture
This PR begins landing a FastGen-like decomposition, adapted to FastVideo realities:

- `ModelBundle(roles: dict[str, RoleHandle])`
  - Treat every participant as a role key (eg. teacher, student, ...).
  - Each role can carry modules / optimizers / schedulers, and the method decides what it needs per step.
  - Reason: distillation setups vary widely; hard-coding role semantics blocks extensibility.

- `DistillMethod`
  - Owns the algorithm: defines `single_train_step()` and selects which optimizers/schedulers should step at a given iteration (supports `generator_update_interval`, multi-optimizer schedules, etc.).
  - Reason: optimizer/scheduler cadence is part of the algorithm, not the trainer.

- `DistillTrainer`
  - A generic training loop that does grad accumulation, logging, and calls method hooks.
  - Reason: keep the trainer algorithm-agnostic and easy to review.

- `DistillAdapter`
  - The boundary that absorbs pipeline/model differences (batch structure, conditioning/CFG, attention metadata, layout/normalize, and forward-context needs).
  - Reason: FastVideo pipelines differ significantly across models; an adapter is the right place to normalize these differences so methods can remain reusable.

```mermaid
flowchart LR
  %% --- 样式定义区 ---
  classDef config fill:#e3f2fd,stroke:#1565c0,stroke-width:2px,rx:8,ry:8,color:#0d47a1,font-weight:bold
  classDef catalogItem fill:#f5f5f5,stroke:#b0bec5,stroke-width:1px,rx:4,ry:4,color:#546e7a
  classDef selectedModel fill:#fff3e0,stroke:#e65100,stroke-width:2px,rx:6,ry:6,color:#e65100,font-weight:bold
  classDef selectedMethod fill:#e8f5e9,stroke:#2e7d32,stroke-width:2px,rx:6,ry:6,color:#1b5e20,font-weight:bold
  classDef infra fill:#f3e5f5,stroke:#8e24aa,stroke-width:2px,rx:8,ry:8,color:#4a148c,font-weight:bold
  classDef action fill:#ffebee,stroke:#c62828,stroke-width:2px,rx:8,ry:8,color:#c62828,font-weight:bold

  %% --- 配置入口 ---
  C["Config<br/>(select models + method)"]:::config

  %% --- 备选库 (垂直排列) ---
  subgraph Models ["Models (Catalog)"]
    direction TB
    Mo1["Wan 1.3B"]:::selectedModel
    Mo2["Wan 14B"]:::selectedModel
    Mo3["SDXL"]:::catalogItem
    Mo4["Flux"]:::catalogItem
  end

  subgraph Methods ["Methods (Catalog)"]
    direction TB
    Me1["distribution_matching / DMD2"]:::catalogItem
    Me2["knowledge_distillation / KD"]:::catalogItem
    Me3["fine_tuning / SFT"]:::catalogItem
    Me4["consistency_model / CM"]:::selectedMethod
  end

  C --> Models
  C --> Methods

  %% --- 下游组装流水线 ---
  %% 1. 从目录中的模型直接指向专属的 Adapter
  Mo1 --> AdapterS["ModelAdapter<br/>(WanAdapter)"]:::infra
  Mo2 --> AdapterT["ModelAdapter<br/>(WanAdapter)"]:::infra
  
  %% 从目录中直接引出实例化的 Method
  Me4 --> PickMethod["DistillMethod<br/>(CM Instance)"]:::selectedMethod

  %% 2. 分别经过 ModelBundle 打包并分配角色
  AdapterS --> BundleS["ModelBundle<br/>(role: student)"]:::infra
  AdapterT --> BundleT["ModelBundle<br/>(role: teacher)"]:::infra

  %% 3. Bundle 注入到 Method，Trainer 只接收 Method
  BundleS --> PickMethod
  BundleT --> PickMethod
  
  PickMethod --> Trainer["DistillTrainer"]:::infra
  Trainer --> Out["Train / Run"]:::action

  %% --- 子图背景框样式 ---
  style Models fill:none,stroke:#ffb74d,stroke-width:2px,stroke-dasharray: 5 5,color:#e65100
  style Methods fill:none,stroke:#aed581,stroke-width:2px,stroke-dasharray: 5 5,color:#33691e
```

Phase 0 intentionally uses a *pipeline-backed adapter* (reusing legacy pipeline helpers) to de-risk initial bring-up. Later phases will progressively move those helpers into a true adapter to fully remove coupling.

### ✅ Done in this PR (Phase 0)
- Add an additive distillation scaffold (does not modify existing distillation pipelines):
  - `fastvideo/distillation/bundle.py`: `ModelBundle` / `RoleHandle`
  - `fastvideo/distillation/methods/base.py`: `DistillMethod` base + optimizer/scheduler stepping interface
  - `fastvideo/distillation/trainer.py`: `DistillTrainer` (grad-accum training loop)
  - `fastvideo/distillation/adapters/base.py`: `DistillAdapter` base
- Bring up Wan DMD2 on the new scaffold:
  - `fastvideo/distillation/adapters/wan.py`: `WanPipelineAdapter` (pipeline-backed)
  - `fastvideo/distillation/methods/wan_dmd2.py`: `WanDMD2Method` (student/teacher/critic, multi-optimizer schedule)
  - `fastvideo/training/wan_distillation_v2.py`: a new v2 entrypoint to run via Method/Trainer without touching legacy scripts
- Remove one critical hidden coupling:
  - Initialize negative/unconditional conditioning explicitly at train start (no longer depends on validation side effects).
- Add a minimal validation hook for smoke signals:
  - `DistillTrainer` calls `method.log_validation(step)` at start and every `--validation_steps` (Wan implementation reuses legacy `_log_validation`).
- Add an 8-step DMD2 example script:
  - `examples/distillation/phase0/distill_wan2.1_t2v_1.3B_dmd2_8steps.sh`
- Add a CPU-only unit test:
  - `fastvideo/tests/distillation/test_phase0_schedule.py` ensures optimizer/scheduler stepping aligns with `generator_update_interval`.

### ❌ Not done yet (explicit follow-ups)
- Extract a generic `DMD2Method` (algorithm-only) and remove the “Wan” prefix from the method by pushing model specifics into the adapter.
- Evolve `WanPipelineAdapter -> WanAdapter` by eliminating dependencies on legacy pipeline private helpers.
- Proper checkpoint save/resume wiring for the v2 path.
- A more general validation abstraction so each method doesn’t need to implement its own validation logic.
- Additional distillation methods (e.g., self-forcing) and adapters for models beyond Wan.

## Tests / Evidence
- `pytest -vs fastvideo/tests/distillation/test_phase0_schedule.py`
- Manual GPU run of Wan2.* DMD2 shows a strong positive signal: validation videos start blurry early and become progressively clearer and higher quality over training.

A test run: https://wandb.ai/alexzms-ucsd/phase0_wan_dmd2_8steps_wansyn/runs/cpw05inn?nw=nwuseralexzms

## Feedback and suggestions are highly welcome!

### CODE REVIEW

**Comment by jzhang38** [comment]
@alexzms Please run precommit
